File: src/dere_ambient/triggers/collector.py
"""Collect curiosity triggers and store them in the work queue."""

from __future__ import annotations

from datetime import UTC, datetime, timedelta
from typing import Any, TYPE_CHECKING

from loguru import logger
from sqlalchemy import func, select

from dere_shared.models import ProjectTask, ProjectTaskStatus

from .corrections import detect_correction
from .emotions import detect_emotional_peak
from .entities import detect_unfamiliar_entities
from .knowledge_gap import detect_knowledge_gap
from .priority import compute_curiosity_priority
from .types import CuriositySignal
from .unfinished_thread import detect_unfinished_thread

if TYPE_CHECKING:
    from collections.abc import Callable

    from sqlalchemy.ext.asyncio import AsyncSession


async def process_curiosity_triggers(
    *,
    prompt: str,
    session_id: int,
    conversation_id: int,
    user_id: str | None,
    working_dir: str,
    personality: str | None,
    speaker_name: str | None,
    is_command: bool,
    message_type: str = "user",
    kg_nodes: list[Any] | None,
    session_factory: Callable[[], AsyncSession],
) -> int:
    text = prompt.strip()
    if not text:
        return 0

    if message_type == "user" and is_command:
        return 0

    if len(text) < 6:
        return 0

    async with session_factory() as db:
        signals: list[CuriositySignal] = []
        if message_type == "user":
            previous_assistant = await _get_previous_assistant_message(
                db,
                session_id=session_id,
                conversation_id=conversation_id,
            )
            signals.extend(
                detect_unfamiliar_entities(
                    prompt=text,
                    nodes=kg_nodes,
                    speaker_name=speaker_name,
                    personality=personality,
                )
            )

            correction = detect_correction(
                prompt=text, previous_assistant=previous_assistant
            )
            if correction:
                signals.append(correction)

            emotional = detect_emotional_peak(prompt=text)
            if emotional:
                signals.append(emotional)

            unfinished = detect_unfinished_thread(
                prompt=text,
                previous_assistant=previous_assistant,
            )
            if unfinished:
                signals.append(unfinished)
        elif message_type == "assistant":
            previous_user = await _get_previous_user_message(
                db,
                session_id=session_id,
                conversation_id=conversation_id,
            )
            knowledge_gap = detect_knowledge_gap(
                prompt=text,
                previous_user=previous_user,
            )
            if knowledge_gap:
                signals.append(knowledge_gap)

        if not signals:
            return 0

        await _enforce_backlog_limits(db, user_id=user_id)

        created = 0
        seen_topics: set[str] = set()
        for signal in signals:
            normalized = _normalize_topic(signal.topic)
            if normalized in seen_topics:
                continue
            seen_topics.add(normalized)

            logger.info(
                "Curiosity trigger: type={} topic={} reason={}",
                signal.curiosity_type,
                signal.topic,
                signal.trigger_reason,
            )
            created += await _upsert_curiosity_task(
                db,
                signal,
                working_dir=working_dir,
                user_id=user_id,
                conversation_id=conversation_id,
            )

        await db.commit()
        if created:
            logger.info(
                "Curiosity triggers stored: created={} total_signals={}",
                created,
                len(signals),
            )
        return created


async def _get_previous_assistant_message(
    db: AsyncSession,
    *,
    session_id: int,
    conversation_id: int,
) -> str | None:
    from dere_shared.models import Conversation

    stmt = (
        select(Conversation.prompt)
        .where(Conversation.session_id == session_id)
        .where(Conversation.message_type == "assistant")
        .where(Conversation.id < conversation_id)
        .order_by(Conversation.id.desc())
        .limit(1)
    )
    result = await db.execute(stmt)
    return result.scalar_one_or_none()


async def _get_previous_user_message(
    db: AsyncSession,
    *,
    session_id: int,
    conversation_id: int,
) -> str | None:
    from dere_shared.models import Conversation

    stmt = (
        select(Conversation.prompt)
        .where(Conversation.session_id == session_id)
        .where(Conversation.message_type == "user")
        .where(Conversation.id < conversation_id)
        .order_by(Conversation.id.desc())
        .limit(1)
    )
    result = await db.execute(stmt)
    return result.scalar_one_or_none()


async def _upsert_curiosity_task(
    db: AsyncSession,
    signal: CuriositySignal,
    *,
    working_dir: str,
    user_id: str | None,
    conversation_id: int,
) -> int:
    normalized_topic = _normalize_topic(signal.topic)
    stmt = (
        select(ProjectTask)
        .where(ProjectTask.task_type == "curiosity")
        .where(func.lower(ProjectTask.title) == normalized_topic)
        .limit(1)
    )
    result = await db.execute(stmt)
    existing = result.scalar_one_or_none()

    now = datetime.now(UTC)

    if existing and existing.status in {
        ProjectTaskStatus.DONE.value,
        ProjectTaskStatus.CANCELLED.value,
    }:
        existing = None

    if existing:
        extra = dict(existing.extra or {})
        trigger_count = int(extra.get("trigger_count") or 0) + 1
        exploration_count = int(extra.get("exploration_count") or 0)
        recency = _recency_factor(existing, signal)
        priority, factors = compute_curiosity_priority(
            signal,
            exploration_count=exploration_count,
            recency=recency,
        )
        repeat_bonus = min(0.2, 0.05 * trigger_count)
        priority = min(1.0, priority + repeat_bonus)
        factors["repeat_bonus"] = repeat_bonus

        existing.priority = max(existing.priority, int(priority * 100))
        existing.updated_at = now
        existing.extra = _merge_extra(
            extra,
            signal,
            priority_factors=factors,
            trigger_count=trigger_count,
            conversation_id=conversation_id,
            user_id=user_id,
            now=now,
        )
        logger.info(
            "Updated curiosity task {}: type={} priority={} triggers={}",
            existing.id,
            signal.curiosity_type,
            existing.priority,
            trigger_count,
        )
        return 0

    priority, factors = compute_curiosity_priority(signal)
    task = ProjectTask(
        working_dir=working_dir,
        title=signal.topic,
        description=f"Curiosity trigger: {signal.trigger_reason}",
        task_type="curiosity",
        priority=int(priority * 100),
        status=ProjectTaskStatus.READY.value,
        extra=_merge_extra(
            {},
            signal,
            priority_factors=factors,
            trigger_count=1,
            conversation_id=conversation_id,
            user_id=user_id,
            now=now,
        ),
        created_at=now,
        updated_at=now,
    )
    db.add(task)
    logger.info(
        "Created curiosity task: type={} topic={} priority={}",
        signal.curiosity_type,
        signal.topic,
        task.priority,
    )
    return 1


async def _enforce_backlog_limits(db: AsyncSession, *, user_id: str | None) -> None:
    max_pending = 100
    max_per_type = 25
    prune_threshold = 0.15

    pending_statuses = {
        ProjectTaskStatus.BACKLOG.value,
        ProjectTaskStatus.READY.value,
        ProjectTaskStatus.BLOCKED.value,
    }

    query = select(ProjectTask).where(ProjectTask.task_type == "curiosity")
    query = query.where(ProjectTask.status.in_(pending_statuses))
    if user_id:
        query = query.where(ProjectTask.extra["user_id"].astext == user_id)

    result = await db.execute(query)
    tasks = list(result.scalars().all())
    if not tasks:
        return

    now = datetime.now(UTC)
    to_cancel: set[int] = set()

    for task in tasks:
        if _should_prune_task(task, now, prune_threshold):
            to_cancel.add(task.id)

    remaining = [t for t in tasks if t.id not in to_cancel]
    if len(remaining) > max_pending:
        overflow = len(remaining) - max_pending
        for task in _lowest_priority(remaining)[:overflow]:
            to_cancel.add(task.id)
            remaining.remove(task)

    by_type: dict[str, list[ProjectTask]] = {}
    for task in remaining:
        curiosity_type = _task_curiosity_type(task)
        by_type.setdefault(curiosity_type, []).append(task)

    for curiosity_type, bucket in by_type.items():
        if len(bucket) <= max_per_type:
            continue
        overflow = len(bucket) - max_per_type
        for task in _lowest_priority(bucket)[:overflow]:
            to_cancel.add(task.id)

    for task in tasks:
        if task.id not in to_cancel:
            continue
        task.status = ProjectTaskStatus.CANCELLED.value
        task.updated_at = now
        task.last_error = "pruned by backlog limits"
        extra = dict(task.extra or {})
        extra["pruned_at"] = now.isoformat()
        extra.setdefault("pruned_reason", "backlog_limits")
        task.extra = extra


def _should_prune_task(
    task: ProjectTask,
    now: datetime,
    prune_threshold: float,
) -> bool:
    curiosity_type = _task_curiosity_type(task)
    ttl_days = 7 if curiosity_type == "correction" else 14
    cutoff = now - timedelta(days=ttl_days)

    last_triggered = _parse_iso_datetime(
        (task.extra or {}).get("last_triggered_at")
    )
    effective_time = last_triggered or task.created_at
    if effective_time and effective_time < cutoff:
        return True

    priority = (task.priority or 0) / 100
    return priority < prune_threshold


def _lowest_priority(tasks: list[ProjectTask]) -> list[ProjectTask]:
    return sorted(
        tasks,
        key=lambda task: (task.priority or 0, task.created_at or datetime.min),
    )


def _task_curiosity_type(task: ProjectTask) -> str:
    extra = task.extra or {}
    return str(extra.get("curiosity_type") or "unknown")


def _recency_factor(task: ProjectTask, signal: CuriositySignal) -> float:
    base = 1.0
    ttl_days = 7 if signal.curiosity_type == "correction" else 14
    last_triggered = _parse_iso_datetime(
        (task.extra or {}).get("last_triggered_at")
    )
    effective_time = last_triggered or task.created_at
    if not effective_time:
        return base
    age_days = max(0.0, (datetime.now(UTC) - effective_time).total_seconds() / 86400)
    return max(0.0, 1.0 - (age_days / ttl_days))


def _parse_iso_datetime(value: Any) -> datetime | None:
    if not value:
        return None
    if isinstance(value, datetime):
        return value
    try:
        return datetime.fromisoformat(str(value))
    except Exception:
        return None


def _merge_extra(
    base: dict[str, Any],
    signal: CuriositySignal,
    *,
    priority_factors: dict[str, float],
    trigger_count: int,
    conversation_id: int,
    user_id: str | None,
    now: datetime,
) -> dict[str, Any]:
    extra = dict(base)
    extra.update(
        {
            "curiosity_type": signal.curiosity_type,
            "source_context": signal.source_context,
            "trigger_reason": signal.trigger_reason,
            "priority_factors": priority_factors,
            "trigger_count": trigger_count,
            "last_triggered_at": now.isoformat(),
            "user_id": user_id,
            "conversation_id": conversation_id,
        }
    )

    extra.setdefault("findings", [])
    extra.setdefault("exploration_count", 0)
    extra.setdefault("last_explored_at", None)
    extra.setdefault("satisfaction_level", 0.0)

    return extra


def _normalize_topic(topic: str) -> str:
    return topic.strip().lower()

File: packages/daemon/src/ambient-triggers/collector.ts
import { sql, type Kysely } from "kysely";

import type { Database } from "../db-types.js";
import { detectCorrection } from "./corrections.js";
import { detectEmotionalPeak } from "./emotions.js";
import { detectUnfamiliarEntities, type EntityNodeLike } from "./entities.js";
import { detectKnowledgeGap } from "./knowledge-gap.js";
import { computeCuriosityPriority } from "./priority.js";
import type { CuriositySignal } from "./types.js";
import { log } from "../logger.js";

const STATUS = {
  BACKLOG: "backlog",
  READY: "ready",
  BLOCKED: "blocked",
  DONE: "done",
  CANCELLED: "cancelled",
  IN_PROGRESS: "in_progress",
} as const;

type DbTask = {
  id: number;
  title: string;
  status: string;
  priority: number | null;
  created_at: Date | null;
  updated_at: Date | null;
  extra: Record<string, unknown> | null;
};

function toJsonRecord(value: unknown): Record<string, unknown> | null {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    return null;
  }
  return value as Record<string, unknown>;
}

function normalizeTask(task: {
  id: number;
  title: string;
  status: string;
  priority: number | null;
  created_at: Date | null;
  updated_at: Date | null;
  extra: unknown;
}): DbTask {
  return { ...task, extra: toJsonRecord(task.extra) };
}

export interface ProcessCuriosityOptions {
  db: Kysely<Database>;
  prompt: string;
  sessionId: number;
  conversationId: number;
  userId: string | null;
  workingDir: string;
  personality: string | null;
  speakerName: string | null;
  isCommand: boolean;
  messageType: string;
  kgNodes: EntityNodeLike[] | null;
}

export async function processCuriosityTriggers(options: ProcessCuriosityOptions): Promise<number> {
  const {
    db,
    prompt,
    sessionId,
    conversationId,
    userId,
    workingDir,
    personality,
    speakerName,
    isCommand,
    messageType,
    kgNodes,
  } = options;

  const text = prompt.trim();
  if (!text || text.length < 6) {
    return 0;
  }
  if (messageType === "user" && isCommand) {
    return 0;
  }

  const signals: CuriositySignal[] = [];
  if (messageType === "user") {
    const previousAssistant = await getPreviousMessage(db, {
      sessionId,
      conversationId,
      role: "assistant",
    });

    detectUnfamiliarEntities({
      prompt: text,
      nodes: kgNodes,
      speakerName,
      personality,
    }).forEach((signal) => {
      signals.push({
        curiosity_type: "unfamiliar_entity",
        topic: signal.topic,
        source_context: signal.sourceContext,
        trigger_reason: signal.triggerReason,
        user_interest: 0.4,
        knowledge_gap: 0,
        metadata: {},
      });
    });

    const correction = detectCorrection(text, previousAssistant);
    if (correction) {
      signals.push({
        curiosity_type: "correction",
        topic: correction.topic,
        source_context: correction.sourceContext,
        trigger_reason: correction.triggerReason,
        user_interest: 0.7,
        knowledge_gap: 0,
        metadata: {},
      });
    }

    const emotional = detectEmotionalPeak(text);
    if (emotional) {
      signals.push({
        curiosity_type: "emotional_peak",
        topic: emotional.topic,
        source_context: emotional.sourceContext,
        trigger_reason: emotional.triggerReason,
        user_interest: emotional.userInterest,
        knowledge_gap: 0,
        metadata: { intensity: emotional.intensity },
      });
    }
  } else if (messageType === "assistant") {
    const previousUser = await getPreviousMessage(db, {
      sessionId,
      conversationId,
      role: "user",
    });
    const knowledgeGap = detectKnowledgeGap(text, previousUser);
    if (knowledgeGap) {
      signals.push({
        curiosity_type: "knowledge_gap",
        topic: knowledgeGap.topic,
        source_context: knowledgeGap.sourceContext,
        trigger_reason: knowledgeGap.triggerReason,
        user_interest: 0.4,
        knowledge_gap: 0.8,
        metadata: {},
      });
    }
  }

  if (signals.length === 0) {
    return 0;
  }

  return await db.transaction().execute(async (trx) => {
    await enforceBacklogLimits(trx, userId);

    let created = 0;
    const seenTopics = new Set<string>();
    for (const signal of signals) {
      const normalized = normalizeTopic(signal.topic);
      if (seenTopics.has(normalized)) {
        continue;
      }
      seenTopics.add(normalized);

      created += await upsertCuriosityTask(trx, signal, {
        workingDir,
        userId,
        conversationId,
      });
    }

    if (created > 0) {
      log.ambient.info("Curiosity triggers stored", { created, total: signals.length });
    }

    return created;
  });
}

async function getPreviousMessage(
  db: Kysely<Database>,
  options: { sessionId: number; conversationId: number; role: string },
): Promise<string | null> {
  const row = await db
    .selectFrom("conversations")
    .select(["prompt"])
    .where("session_id", "=", options.sessionId)
    .where("message_type", "=", options.role)
    .where("id", "<", options.conversationId)
    .orderBy("id", "desc")
    .limit(1)
    .executeTakeFirst();

  return row?.prompt ?? null;
}

async function upsertCuriosityTask(
  db: Kysely<Database>,
  signal: CuriositySignal,
  options: { workingDir: string; userId: string | null; conversationId: number },
): Promise<number> {
  const normalizedTopic = normalizeTopic(signal.topic);
  const existingRaw = await db
    .selectFrom("project_tasks")
    .select(["id", "title", "status", "priority", "created_at", "updated_at", "extra"])
    .where("task_type", "=", "curiosity")
    .where(sql<string>`lower(title)`, "=", normalizedTopic)
    .limit(1)
    .executeTakeFirst();
  const existing = existingRaw ? normalizeTask(existingRaw) : null;

  const now = new Date();

  if (existing && (existing.status === STATUS.DONE || existing.status === STATUS.CANCELLED)) {
    return 0;
  }

  if (existing) {
    const extra = normalizeExtra(existing.extra);
    const triggerCount = Number(extra.trigger_count ?? 0) + 1;
    const explorationCount = Number(extra.exploration_count ?? 0);
    const recency = recencyFactor(existing, signal);
    const { score, factors } = computeCuriosityPriority({
      signal,
      explorationCount,
      recency,
    });
    const repeatBonus = Math.min(0.2, 0.05 * triggerCount);
    const priority = Math.min(1.0, score + repeatBonus);
    factors.repeat_bonus = repeatBonus;

    const updatedExtra = mergeExtra(extra, signal, {
      priorityFactors: factors,
      triggerCount,
      conversationId: options.conversationId,
      userId: options.userId,
      now,
    });

    await db
      .updateTable("project_tasks")
      .set({
        priority: Math.max(existing.priority ?? 0, Math.floor(priority * 100)),
        updated_at: now,
        extra: updatedExtra,
      })
      .where("id", "=", existing.id)
      .execute();

    log.ambient.info("Updated curiosity task", {
      taskId: existing.id,
      type: signal.curiosity_type,
      priority: Math.floor(priority * 100),
      triggers: triggerCount,
    });
    return 0;
  }

  const { score, factors } = computeCuriosityPriority({ signal });
  const extra = mergeExtra({}, signal, {
    priorityFactors: factors,
    triggerCount: 1,
    conversationId: options.conversationId,
    userId: options.userId,
    now,
  });

  await db
    .insertInto("project_tasks")
    .values({
      working_dir: options.workingDir,
      title: signal.topic,
      description: `Curiosity trigger: ${signal.trigger_reason}`,
      task_type: "curiosity",
      priority: Math.floor(score * 100),
      status: STATUS.READY,
      extra,
      created_at: now,
      updated_at: now,
      started_at: null,
      completed_at: null,
      acceptance_criteria: null,
      context_summary: null,
      scope_paths: null,
      required_tools: null,
      tags: null,
      estimated_effort: null,
      claimed_by_session_id: null,
      claimed_by_agent_id: null,
      claimed_at: null,
      attempt_count: 0,
      blocked_by: null,
      related_task_ids: null,
      created_by_session_id: null,
      created_by_agent_id: null,
      discovered_from_task_id: null,
      discovery_reason: null,
      outcome: null,
      completion_notes: null,
      files_changed: null,
      follow_up_task_ids: null,
      last_error: null,
    })
    .execute();

  log.ambient.info("Created curiosity task", {
    type: signal.curiosity_type,
    topic: signal.topic,
    priority: Math.floor(score * 100),
  });

  return 1;
}

async function enforceBacklogLimits(db: Kysely<Database>, userId: string | null): Promise<void> {
  const maxPending = 100;
  const maxPerType = 25;
  const pruneThreshold = 0.15;
  const pendingStatuses = [STATUS.BACKLOG, STATUS.READY, STATUS.BLOCKED];

  let query = db
    .selectFrom("project_tasks")
    .select(["id", "title", "status", "priority", "created_at", "updated_at", "extra"])
    .where("task_type", "=", "curiosity")
    .where("status", "in", pendingStatuses);

  if (userId) {
    query = query.where(sql<boolean>`extra->>'user_id' = ${userId}`);
  }

  const tasks = (await query.execute()).map(normalizeTask);
  if (tasks.length === 0) {
    return;
  }

  const now = new Date();
  const toCancel = new Set<number>();

  for (const task of tasks) {
    if (shouldPruneTask(task, now, pruneThreshold)) {
      toCancel.add(task.id);
    }
  }

  let remaining = tasks.filter((task) => !toCancel.has(task.id));
  if (remaining.length > maxPending) {
    const overflow = remaining.length - maxPending;
    const lowest = lowestPriority(remaining).slice(0, overflow);
    for (const task of lowest) {
      toCancel.add(task.id);
      remaining = remaining.filter((item) => item.id !== task.id);
    }
  }

  const byType = new Map<string, DbTask[]>();
  for (const task of remaining) {
    const curiosityType = taskCuriosityType(task);
    const bucket = byType.get(curiosityType) ?? [];
    bucket.push(task);
    byType.set(curiosityType, bucket);
  }

  for (const bucket of byType.values()) {
    if (bucket.length <= maxPerType) {
      continue;
    }
    const overflow = bucket.length - maxPerType;
    for (const task of lowestPriority(bucket).slice(0, overflow)) {
      toCancel.add(task.id);
    }
  }

  if (toCancel.size === 0) {
    return;
  }

  for (const task of tasks) {
    if (!toCancel.has(task.id)) {
      continue;
    }
    const extra = normalizeExtra(task.extra);
    extra.pruned_at = now.toISOString();
    if (!extra.pruned_reason) {
      extra.pruned_reason = "backlog_limits";
    }
    await db
      .updateTable("project_tasks")
      .set({
        status: STATUS.CANCELLED,
        updated_at: now,
        last_error: "pruned by backlog limits",
        extra,
      })
      .where("id", "=", task.id)
      .execute();
  }
}

function shouldPruneTask(task: DbTask, now: Date, pruneThreshold: number): boolean {
  const curiosityType = taskCuriosityType(task);
  const ttlDays = curiosityType === "correction" ? 7 : 14;
  const cutoff = new Date(now.getTime() - ttlDays * 24 * 60 * 60 * 1000);

  const lastTriggered = parseIsoDate(task.extra?.last_triggered_at);
  const effectiveTime = lastTriggered ?? task.created_at;
  if (effectiveTime && effectiveTime < cutoff) {
    return true;
  }

  const priority = (task.priority ?? 0) / 100;
  return priority < pruneThreshold;
}

function lowestPriority(tasks: DbTask[]): DbTask[] {
  return [...tasks].sort((a, b) => {
    const ap = a.priority ?? 0;
    const bp = b.priority ?? 0;
    if (ap !== bp) {
      return ap - bp;
    }
    const aCreated = a.created_at?.getTime() ?? 0;
    const bCreated = b.created_at?.getTime() ?? 0;
    return aCreated - bCreated;
  });
}

function taskCuriosityType(task: DbTask): string {
  return String(task.extra?.curiosity_type ?? "unknown");
}

function recencyFactor(task: DbTask, signal: CuriositySignal): number {
  const ttlDays = signal.curiosity_type === "correction" ? 7 : 14;
  const lastTriggered = parseIsoDate(task.extra?.last_triggered_at);
  const effectiveTime = lastTriggered ?? task.created_at;
  if (!effectiveTime) {
    return 1.0;
  }
  const ageDays = Math.max(0, (Date.now() - effectiveTime.getTime()) / 86400000);
  return Math.max(0, 1.0 - ageDays / ttlDays);
}

function parseIsoDate(value: unknown): Date | null {
  if (!value) {
    return null;
  }
  if (value instanceof Date) {
    return value;
  }
  const parsed = new Date(String(value));
  return Number.isNaN(parsed.getTime()) ? null : parsed;
}

function mergeExtra(
  base: Record<string, unknown>,
  signal: CuriositySignal,
  options: {
    priorityFactors: Record<string, number>;
    triggerCount: number;
    conversationId: number;
    userId: string | null;
    now: Date;
  },
): Record<string, unknown> {
  const extra: Record<string, unknown> = { ...base };
  extra.curiosity_type = signal.curiosity_type;
  extra.source_context = signal.source_context;
  extra.trigger_reason = signal.trigger_reason;
  extra.priority_factors = options.priorityFactors;
  extra.trigger_count = options.triggerCount;
  extra.last_triggered_at = options.now.toISOString();
  extra.user_id = options.userId;
  extra.conversation_id = options.conversationId;

  if (extra.findings === undefined) {
    extra.findings = [];
  }
  if (extra.exploration_count === undefined) {
    extra.exploration_count = 0;
  }
  if (extra.last_explored_at === undefined) {
    extra.last_explored_at = null;
  }
  if (extra.satisfaction_level === undefined) {
    extra.satisfaction_level = 0;
  }

  return extra;
}

function normalizeExtra(extra: Record<string, unknown> | null): Record<string, unknown> {
  return extra ? { ...extra } : {};
}

function normalizeTopic(topic: string): string {
  return topic.trim().toLowerCase();
}
